### ISSUE TITLE
client-audit: error reading packet

### ISSUE THREAD

**Issue by Izorkin** [jtesta/ssh-audit#56]
Not working function - cleint-audit. Error:
```
ssh-audit --client-audit --timeout=300 --verbose
# general
(gen) banner: SSH-2.0-OpenSSH_8.0
(gen) software: OpenSSH 8.0

[exception] error reading packet (empty)
```
Check with python 3.7 and python 3.8.

**Comment by jugmac00** [comment]
FWIW - I cannot reproduce this error with the current packaged version 2.2.0 on Ubuntu 18.04.

```
ssh-audit --client-audit --timeout=300 --verbose
# general
(gen) client IP: 127.0.0.1
(gen) banner: SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
(gen) software: OpenSSH 7.6p1
(gen) compression: enabled (zlib@openssh.com, zlib)
...
```

I am curious what @jtesta thinks about this problem.

Disclaimer: I am no maintainer, just a user.

**Comment by Izorkin** [comment]
Possible error due to OpenSSH version?
Used 8.2.

**Comment by jtesta** [comment]
I can't reproduce this error with the `dev` branch:
```
$ ssh -V
OpenSSH_8.2p1 Ubuntu-4ubuntu0.1, OpenSSL 1.1.1f  31 Mar 2020
$ python3 -V
Python 3.8.2
$ python3 ssh-audit.py --client-audit --timeout=300 --verbose
# general
(gen) client IP: 127.0.0.1
(gen) banner: SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.1
(gen) software: OpenSSH 8.2p1
(gen) compression: enabled (zlib@openssh.com, zlib)

# key exchange algorithms
[...]
```

Are you using the `master` branch?  Or a release?

**Comment by Izorkin** [comment]
Yes, error reproduced with the `dev` branch:
```
ssh -V
OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020
```
```
python3 ssh-audit.py --client-audit --timeout=300 --verbose
# general
(gen) banner: SSH-2.0-OpenSSH_8.0
(gen) software: OpenSSH 8.0

[exception] error reading packet (empty)
Time: 0h:00m:08s
```

**Comment by Izorkin** [comment]
Sorry I did not test correctly.
To check, need run the command `ssh 127.0.0.1 -p 2222`, not this command `ssh-audit 127.0.0.1 --port 2222`

**Comment by jugmac00** [comment]
Both `-p` and `--port` should be equivalent.

Also, now I can reproduce the failure.

```
❯ python3.8 ssh-audit.py localhost -p 2222
# general
(gen) banner: SSH-2.0-OpenSSH_8.0
(gen) software: OpenSSH 8.0

[exception] error reading packet (timed out)
```

Please reopen the issue.

**Comment by jtesta** [comment]
I did notice that you were running ssh-audit against itself as a server and a client.

The reason neither side successfully completes is that both the client & server exchange banners, then wait for the other side to send the list of supported algorithms.  Since each side waits for the other, nothing ever happens.  So either the server's timeout is reached, or the client's is.  When a timeout of 300 is set on the server, the client's default timeout of 5 seconds is reached first, so the client terminates the connection and the server prints the `[exception] error reading packet (empty)` message.  On the other hand, if you set the server's timeout to 10, and the client's to 300, you'll find that the server prints `[exception] error reading packet (timed out)`.

I've changed the behavior so that a list of algorithms are sent to the peer before attempting to read the peer's list.  I don't so much care about fixing this weird case of having ssh-audit run against itself, than the possibility of a real SSH peer doing something similar in a real environment.  So perhaps this preemptively fixes a bug.

@jugmac00 : Does [1123ac7](https://github.com/jtesta/ssh-audit/commit/1123ac718c674597ca84c05da3c0cbd4050f39d1) fix this for you?  If not, I'll re-open this issue.

@Izorkin : I appreciate the report!
